src/PrestaShopBundle/Translation/Provider/UserTranslatedCatalogueProvider.php

```diff
@@ -0,0 +1,111 @@
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+declare(strict_types=1);
+
+namespace PrestaShopBundle\Translation\Provider;
+
+use PrestaShopBundle\Translation\Loader\DatabaseTranslationLoader;
+use Symfony\Component\Translation\MessageCatalogue;
+
+/**
+ * Gets catalogue translated by the user himself ans stored in the database.
+ */
+class UserTranslatedCatalogueProvider
+{
+    /**
+     * @var DatabaseTranslationLoader
+     */
+    private $databaseTranslationReader;
+
+    /**
+     * @var array
+     */
+    private $translationDomains;
+
+    /**
+     * @var string|null
+     */
+    private $themeName;
+
+    /**
+     * You will need to give theme if you want only the translations linked to a specific theme.
+     * If not given, the translations returns will be the ones with 'theme IS NULL'
+     *
+     * @param DatabaseTranslationLoader $databaseTranslationReader
+     * @param array $translationDomains
+     * @param string|null $themeName
+     */
+    public function __construct(
+        DatabaseTranslationLoader $databaseTranslationReader,
+        array $translationDomains,
+        ?string $themeName = null
+    ) {
+        if (!$this->assertIsArrayOfString($translationDomains)) {
+            throw new \InvalidArgumentException('Given translation domains are invalid. An array of strings was expected.');
+        }
+
+        $this->databaseTranslationReader = $databaseTranslationReader;
+        $this->translationDomains = $translationDomains;
+        $this->themeName = $themeName;
+    }
+
+    /**
+     * Returns the translation catalogue for the provided locale
+     *
+     * @param string $locale
+     *
+     * @return MessageCatalogue
+     */
+    public function getCatalogue(string $locale): MessageCatalogue
+    {
+        $catalogue = new MessageCatalogue($locale);
+
+        foreach ($this->translationDomains as $translationDomain) {
+            $domainCatalogue = $this->databaseTranslationReader->load(
+                null,
+                $locale,
+                $translationDomain,
+                $this->themeName
+            );
+
+            $catalogue->addCatalogue($domainCatalogue);
+        }
+
+        return $catalogue;
+    }
+
+    /**
+     * Validate if an array only have strings in it.
+     *
+     * @param array $array
+     *
+     * @return bool
+     */
+    private function assertIsArrayOfString(array $array): bool
+    {
+        return count($array) === count(array_filter($array, function ($element) { return is_string($element); }));
```

Comment:
```suggestion
        return count($array) === count(array_filter($array, 'is_string'));
```

But this check is a duplicate. Why not using a ProviderAbstract class to prevent it?

Reply:
Beware, it's not because 2 methods are identical that they should be reunited inside a common abstraction 😉  .

Do we really need to store in a single place, a single SOT, how to check whether a given array contains only strings?

See also https://dev.to/wuz/stop-trying-to-be-so-dry-instead-write-everything-twice-wet-5g33